File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    
    public int solution(int[][] triangle) {
        int answer = 0;
        int N = triangle.length;
        
        int[][] Dy = new int[N][N];
        Dy[0][0] = triangle[0][0];
        for(int i = 1; i < N; i++) {
            for(int j = 0; j < triangle[i].length; j++) {
                if(j == 0) {
                    Dy[i][j] = Dy[i - 1][j] + triangle[i][j];
                    continue;
                }
                if(j == triangle[i].length - 1) {
                    Dy[i][j] = Dy[i - 1][j - 1] + triangle[i][j];
                    continue;
                }
                Dy[i][j] = Math.max(Dy[i - 1][j - 1], Dy[i - 1][j]) + triangle[i][j];
            }
        }
        
        for(int i = 0; i < N; i++) {
            answer = Math.max(answer, Dy[N - 1][i]);
        }
        
        return answer;
    }
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    let answer = 0;
    const N = triangle.length;
    
    let arr = new Array();
    for(let i = 0; i < N; i++) {
        arr[i] = [];
        for(let j = 0; j < N; j++) {
            arr[i][j] = 0;
        }
    }
    
    arr[0][0] = triangle[0][0];
    
    for(let i = 1; i < N; i++) {
        for(let j = 0; j < triangle[i].length; j++) {
            if(j === 0) {
                arr[i][j] = arr[i - 1][j] + triangle[i][j];
            } else if(j === triangle[i].length - 1) {
                arr[i][j] = arr[i - 1][j - 1] + triangle[i][j]; 
            } else {
                arr[i][j] = Math.max(arr[i - 1][j - 1], arr[i - 1][j]) + triangle[i][j];
            }
        }
    }
    
    for(let i = 0; i < N; i++) {
        answer = Math.max(arr[N - 1][i], answer);
    }
    
    return answer;
}